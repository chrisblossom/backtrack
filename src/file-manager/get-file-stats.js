/* @flow */

import type { ParsedFiles, FileManagerStats } from '../types.js';

function getFileStats(parsedFiles: ParsedFiles): FileManagerStats {
    const fileHashes = parsedFiles.dest.files.reduce((acc, destFile) => {
        const destFileIndex = parsedFiles.dest.files.indexOf(destFile);

        /**
         * Use source file's hash to allow for 'allowChanges' option and guard against file system errors
         *
         * destHash will equal srcHash because the copy_files task is done prior to stats
         */
        const srcFile = parsedFiles.src.files[destFileIndex];
        const srcHash = parsedFiles.src.hash[srcFile];

        return {
            ...acc,
            [destFile]: srcHash,
        };
    }, {});

    const stats = {};
    if (Object.keys(fileHashes).length !== 0) {
        stats.files = fileHashes;
    }

    if (parsedFiles.makeDirs.length !== 0) {
        stats.directories = parsedFiles.makeDirs;
    }

    return stats;
}

export { getFileStats };
