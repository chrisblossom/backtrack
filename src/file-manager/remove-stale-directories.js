/* @flow */

import path from 'path';
import { existsSync, readdir } from 'fs';
import del from 'del';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { normalizePathname } from '../utils/normalize-pathname';

import type { ParsedFiles, DirStats } from '../types.js';

function readDir(pathname) {
    return new Promise((resolve, reject) => {
        readdir(pathname, (error, list) => {
            if (error) {
                reject(error);
                return;
            }

            resolve(list);
        });
    });
}

async function shouldDelete(relativeDir, makeDirs = []) {
    async function goUpOne() {
        const upOneDir = path.join(relativeDir, '..');
        if (upOneDir === '.') {
            return;
        }

        await shouldDelete(upOneDir, makeDirs);
    }

    if (makeDirs.includes(relativeDir) === true) {
        return;
    }

    const absolutePath = path.resolve(rootPath, relativeDir);
    const exists = existsSync(absolutePath);
    if (exists === false) {
        await goUpOne();

        return;
    }

    const directoryList = await readDir(absolutePath);
    if (directoryList.length !== 0) {
        log.info(`Directory not empty: ${relativeDir}`);

        return;
    }

    if (path.resolve(relativeDir) === rootPath) {
        return;
    }

    log.info(`Removing directory: ${relativeDir}`);

    await del(absolutePath);

    await goUpOne();
}

async function removeStaleDirectories(
    parsedFiles: ParsedFiles,
    previousStats?: DirStats = [],
) {
    for (const dir of previousStats) {
        const normalizedDir = normalizePathname(dir);

        /**
         * Run sequentially to ensure directories are empty before moving to next
         */
        // eslint-disable-next-line no-await-in-loop
        await shouldDelete(normalizedDir, parsedFiles.makeDirs);
    }
}

export { removeStaleDirectories };
