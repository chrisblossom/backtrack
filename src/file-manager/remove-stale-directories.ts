import path from 'path';
import { existsSync, readdir } from 'fs';
import del from 'del';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { normalizePathname } from '../utils/normalize-pathname';
import { fileIsInsideDir } from '../utils/file-is-inside-dir';

import { ParsedFiles, DirStats } from '../types';

function readDir(pathname: string): Promise<string[]> {
    return new Promise((resolve, reject) => {
        readdir(pathname, (error, list) => {
            if (error) {
                reject(error);
                return;
            }

            resolve(list);
        });
    });
}

async function shouldDelete(relativeDir: string, makeDirs: string[] = []) {
    async function goUpOne() {
        const upOneDir = path.join(relativeDir, '..');
        if (upOneDir === '.') {
            return;
        }

        await shouldDelete(upOneDir, makeDirs);
    }

    if (makeDirs.includes(relativeDir) === true) {
        return;
    }

    const absolutePath = path.resolve(rootPath, relativeDir);
    const exists = existsSync(absolutePath);
    if (exists === false) {
        await goUpOne();

        return;
    }

    const isInsideManagedDir = makeDirs.some((dir) => {
        return fileIsInsideDir(dir, relativeDir);
    });

    if (isInsideManagedDir === true) {
        return;
    }

    const directoryList = await readDir(absolutePath);
    if (directoryList.length !== 0) {
        log.info(`Directory not empty: ${relativeDir}`);

        return;
    }

    if (path.resolve(relativeDir) === rootPath) {
        return;
    }

    log.info(`Removing directory: ${relativeDir}`);

    await del(absolutePath);

    await goUpOne();
}

async function removeStaleDirectories(
    parsedFiles: ParsedFiles,
    previousStats: DirStats = [],
) {
    for (const dir of previousStats) {
        const normalizedDir = normalizePathname(dir);

        /**
         * Run sequentially to ensure directories are empty before moving to next
         */
        // eslint-disable-next-line no-await-in-loop
        await shouldDelete(normalizedDir, parsedFiles.makeDirs);
    }
}

export { removeStaleDirectories };
