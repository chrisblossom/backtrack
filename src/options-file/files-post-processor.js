/* @flow */

import path from 'path';
import { parseFilePath } from '../utils/parse-file-path';
import { normalizePathname } from '../utils/normalize-pathname';
import { toArray } from '../utils/object-utils';

import type { FileManager, ParsedFiles } from '../types.js';

function shouldSkip(pathname, skipFiles) {
    const split = pathname.split(path.sep);

    let currentPath;
    for (const sub of split) {
        currentPath = currentPath ? currentPath + path.sep + sub : sub;

        if (skipFiles.includes(currentPath)) {
            return true;
        }
    }

    return false;
}

function splitSections(copyFiles) {
    return copyFiles.reduce(
        (acc, file) => {
            if (file.src && file.dest) {
                return {
                    ...acc,
                    copyFiles: [...acc.copyFiles, file],
                };
            }

            // $FlowIssue
            const { makeDirs = [], ...options } = file;

            return {
                ...acc,
                options: [...acc.options, options],
                makeDirs: [...acc.makeDirs, ...makeDirs],
            };
        },
        { copyFiles: [], options: [], makeDirs: [] },
    );
}

function parseOptions(options) {
    return options.reduce(
        (acc, fileOptions) => {
            if (fileOptions.skip) {
                toArray(fileOptions.skip).forEach((relativePath) => {
                    const normalizedFile = normalizePathname(relativePath);

                    acc.skip.push(normalizedFile);
                });
            }

            if (typeof fileOptions.allowChanges === 'boolean') {
                acc.allowChangesAll = fileOptions.allowChanges;

                return acc;
            }

            toArray(fileOptions.allowChanges).forEach((file) => {
                const normalizedFile = normalizePathname(file);

                acc.allowChanges.push(normalizedFile);
            });

            return acc;
        },
        {
            allowChangesAll: false,
            allowChanges: [],
            skip: [],
        },
    );
}

function getMakeDirs(makeDirs, files, skipFiles) {
    const filePaths = files
        .map((filePath) => {
            const { dir } = path.parse(filePath);

            return dir;
        })
        .filter(Boolean);

    const result = [...filePaths, ...makeDirs]
        // eslint-disable-next-line arrow-body-style
        .map((dir) => normalizePathname(dir))
        .sort()
        .reduceRight((acc, dir) => {
            /**
             * remove skipped
             */
            const skip = shouldSkip(dir, skipFiles);

            if (skip === true) {
                return acc;
            }

            /**
             * removes redundant directories
             * ['src', 'dist', 'dist/static'] -> ['src', 'dist/static']
             */
            const previous = acc[0];
            if (previous) {
                const trimmedPrevious = previous.substring(0, dir.length);

                if (trimmedPrevious === dir) {
                    return acc;
                }
            }

            return [dir, ...acc];
        }, []);

    return result;
}

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: FileManager,
};

/**
 * Normally this would be a normal processor, but we need to have all results up front
 * to determine what to exclude etc. I'm sure the function could be rewritten but this works as is.
 */
function filesPostProcessor({ value = [] }: Args = {}): ParsedFiles {
    const result = {
        src: {
            files: [],
            absolute: {},
            hash: {},
        },
        dest: {
            files: [],
            absolute: {},
            hash: {},
            allowChanges: {},
        },
        makeDirs: [],
    };

    const sections = splitSections(value);
    const options = parseOptions(sections.options);

    for (const file of sections.copyFiles) {
        const src = parseFilePath(file.src);
        const dest = parseFilePath(file.dest);

        const skip = shouldSkip(dest.relative, options.skip);
        if (skip === false) {
            const srcId = src.relative;
            result.src.files.push(srcId);
            result.src.absolute[srcId] = src.absolute;
            result.src.hash[srcId] = src.hash;

            const destId = dest.relative;
            result.dest.files.push(destId);
            result.dest.absolute[destId] = dest.absolute;
            result.dest.hash[destId] = dest.hash;

            const allowChanges =
                options.allowChanges.includes(destId) || !!file.allowChanges;

            result.dest.allowChanges[destId] =
                options.allowChangesAll || allowChanges;
        }
    }

    result.makeDirs = getMakeDirs(
        sections.makeDirs,
        result.dest.files,
        options.skip,
    );

    return result;
}

export { filesPostProcessor };
