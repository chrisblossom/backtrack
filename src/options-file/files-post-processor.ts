import path from 'path';
import { parseFilePath } from '../utils/parse-file-path';
import { normalizePathname } from '../utils/normalize-pathname';
import { toArray } from '../utils/object-utils';

import {
    FileManager,
    CopyFile,
    isCopyFile,
    CopyFileOptions,
    ParsedFiles,
} from '../types';

function shouldSkip(pathname: string, skipFiles: Array<string>) {
    const split = pathname.split(path.sep);

    let currentPath;
    for (const sub of split) {
        currentPath = currentPath ? currentPath + path.sep + sub : sub;

        if (skipFiles.includes(currentPath)) {
            return true;
        }
    }

    return false;
}

type SplitSections = {
    copyFiles: CopyFile[];
    options: CopyFileOptions[];
    makeDirs: string[];
};

function splitSections(copyFiles: FileManager): SplitSections {
    type SectionsAcc = {
        copyFiles: CopyFile[];
        options: CopyFileOptions[];
        makeDirs: string[];
    };

    const sections = copyFiles.reduceRight(
        (acc: SectionsAcc, file) => {
            if (isCopyFile(file)) {
                const isDuplicate = acc.copyFiles.some((accFile) => {
                    return file.dest === accFile.dest;
                });

                /**
                 * Remove duplicates
                 */
                if (isDuplicate) {
                    return acc;
                }

                return {
                    ...acc,
                    copyFiles: [file, ...acc.copyFiles],
                };
            }

            const { makeDirs = [], ...options } = file;

            return {
                ...acc,
                options: [options, ...acc.options],
                makeDirs: [...makeDirs, ...acc.makeDirs],
            };
        },
        { copyFiles: [], options: [], makeDirs: [] },
    );

    return sections;
}

function parseOptions(options: CopyFileOptions[]) {
    type ParseOptionsAcc = {
        allowChangesAll: boolean;
        ignoreUpdatesAll: boolean;
        allowChanges: string[];
        ignoreUpdates: string[];
        skip: string[];
    };

    const result = options.reduce(
        (acc: ParseOptionsAcc, fileOptions) => {
            if (fileOptions.skip) {
                toArray(fileOptions.skip).forEach((relativePath) => {
                    const normalizedFile = normalizePathname(relativePath);

                    acc.skip.push(normalizedFile);
                });
            }

            if (typeof fileOptions.allowChanges === 'boolean') {
                acc.allowChangesAll = fileOptions.allowChanges;

                return acc;
            }

            toArray(fileOptions.allowChanges).forEach((file) => {
                const normalizedFile = normalizePathname(file);

                acc.allowChanges.push(normalizedFile);
            });

            if (typeof fileOptions.ignoreUpdates === 'boolean') {
                acc.ignoreUpdatesAll = fileOptions.ignoreUpdates;

                return acc;
            }

            toArray(fileOptions.ignoreUpdates).forEach((file) => {
                const normalizedFile = normalizePathname(file);

                acc.ignoreUpdates.push(normalizedFile);
            });

            return acc;
        },
        {
            allowChangesAll: false,
            ignoreUpdatesAll: false,
            allowChanges: [],
            ignoreUpdates: [],
            skip: [],
        },
    );

    return result;
}

function getMakeDirs(makeDirs: string[], files: string[], skipFiles: string[]) {
    const filePaths = files
        .map((filePath) => {
            const { dir } = path.parse(filePath);

            return dir;
        })
        .filter(Boolean);

    const result = [...filePaths, ...makeDirs]
        // eslint-disable-next-line arrow-body-style
        .map((dir) => normalizePathname(dir))
        .sort()
        .reduceRight((acc: string[], dir) => {
            /**
             * remove skipped
             */
            const skip = shouldSkip(dir, skipFiles);

            if (skip === true) {
                return acc;
            }

            /**
             * removes redundant directories
             * ['src', 'dist', 'dist/static'] -> ['src', 'dist/static']
             */
            const previous = acc[0];
            if (previous) {
                const trimmedPrevious = previous.substring(0, dir.length);

                if (trimmedPrevious === dir) {
                    return acc;
                }
            }

            return [dir, ...acc];
        }, []);

    return result;
}

type Args = {
    value?: FileManager;
};

/**
 * Normally this would be a normal processor, but we need to have all results up front
 * to determine what to exclude etc. I'm sure the function could be rewritten but this works as is.
 */
function filesPostProcessor({ value = [] }: Args = {}): ParsedFiles {
    const result: ParsedFiles = {
        src: {
            files: [],
            absolute: {},
            hash: {},
        },
        dest: {
            files: [],
            absolute: {},
            hash: {},
            allowChanges: {},
            ignoreUpdates: {},
        },
        makeDirs: [],
    };

    const sections = splitSections(value);
    const options = parseOptions(sections.options);

    for (const file of sections.copyFiles) {
        const src = parseFilePath(file.src);
        const dest = parseFilePath(file.dest);

        const skip = shouldSkip(dest.relative, options.skip);
        if (skip === false) {
            const srcId = src.relative;
            const destId = dest.relative;

            const ignoreUpdates =
                options.ignoreUpdatesAll ||
                options.ignoreUpdates.includes(destId) ||
                !!file.ignoreUpdates;

            result.src.files.push(srcId);
            result.src.absolute[srcId] = src.absolute;
            result.src.hash[srcId] =
                ignoreUpdates === false ? src.hash : 'ignoreUpdates';

            result.dest.files.push(destId);
            result.dest.absolute[destId] = dest.absolute;
            result.dest.hash[destId] = dest.hash;

            const allowChanges =
                ignoreUpdates ||
                options.allowChanges.includes(destId) ||
                !!file.allowChanges;

            result.dest.ignoreUpdates[destId] = ignoreUpdates;

            result.dest.allowChanges[destId] =
                options.ignoreUpdatesAll ||
                options.allowChangesAll ||
                allowChanges;
        }
    }

    result.makeDirs = getMakeDirs(
        sections.makeDirs,
        result.dest.files,
        options.skip,
    );

    return result;
}

export { filesPostProcessor };
