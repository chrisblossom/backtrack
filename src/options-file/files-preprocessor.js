/* @flow */

import path from 'path';
import { toArray } from '../utils/object-utils';

import type { Files } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: Files,
    dirname: string,
};

function filesPreprocessor({ value, dirname }: Args = {}) {
    const resolveSrc = toArray(value).map((files) => {
        if (typeof files === 'boolean') {
            return files;
        }

        if (!files.src) {
            return files;
        }

        const src = path.resolve(dirname, files.src);

        return {
            ...files,
            src,
        };
    });

    return resolveSrc;
}

export { filesPreprocessor };
