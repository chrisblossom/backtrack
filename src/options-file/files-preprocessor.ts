import path from 'path';
import { toArray } from '../utils/object-utils';

import { isCopyFileOptions, Files } from '../types';

type Args = {
	value?: Files;
	dirname?: string;
};

function filesPreprocessor({ value, dirname = '' }: Args = {}) {
	const resolveSrc = toArray(value).map((files) => {
		if (typeof files === 'boolean') {
			return files;
		}

		if (isCopyFileOptions(files)) {
			return files;
		}

		const src = path.resolve(dirname, files.src);

		return {
			...files,
			src,
		};
	});

	return resolveSrc;
}

export { filesPreprocessor };
