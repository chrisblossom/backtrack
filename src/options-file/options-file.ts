import { loadOptionsFile } from './load-options-file';
import { transformConfig } from './transform-config';
import { rootPath } from '../config/paths';

import { Lifecycles } from '../types';

function optionsFile(dirname: string = rootPath): Lifecycles {
    const baseConfig = loadOptionsFile(dirname);

    const config = transformConfig(baseConfig, dirname);

    return config;
}

export { optionsFile };
