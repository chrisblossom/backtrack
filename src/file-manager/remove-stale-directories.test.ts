import os from 'os';
import path from 'path';
import del from 'del';
import fse from 'fs-extra';
import { realpathSync } from 'fs';
import { getRandomInteger } from '../utils/get-random-number';

const filesPreprocessor = (args: any) =>
	require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesPostProcessor = (args: any) =>
	require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles: any) =>
	require('./get-file-stats').getFileStats(parsedFiles);
const removeStaleDirectories = (parsedFiles: any, previousStats: any) =>
	require('./remove-stale-directories').removeStaleDirectories(
		parsedFiles,
		previousStats,
	);

async function dirInfo(files: any = [], previousFiles: any = []) {
	const dirname = process.cwd();
	const normalizedPrevious = filesPreprocessor({
		value: previousFiles,
		dirname,
	});

	const parsedPreviousFiles = filesPostProcessor({
		value: normalizedPrevious,
		dirname,
	});

	const pending = parsedPreviousFiles.makeDirs.map((dir: any) =>
		fse.ensureDir(dir),
	);

	await Promise.all(pending);

	const previousStats = getFileStats(parsedPreviousFiles).directories;

	const normalizedFiles = filesPreprocessor({ value: files, dirname });

	const parsedFiles = filesPostProcessor({ value: normalizedFiles });

	return {
		previousStats,
		parsedFiles,
	};
}

describe('removeStaleDirectories', () => {
	const cwd = process.cwd();
	let delMock: any;

	const dir = path.resolve(
		realpathSync(os.tmpdir()),
		'backtrack',
		`cwd_${getRandomInteger()}`,
	);

	beforeEach(async () => {
		delMock = require.requireMock('del');
		jest.mock('del', () =>
			jest.fn((dirs) => {
				const delActual = require.requireActual('del');

				return delActual(dirs);
			}),
		);

		jest.mock('../utils/log', () => ({
			warn: jest.fn(),
			info: jest.fn(),
			error: jest.fn(),
			success: jest.fn(),
		}));

		await fse.ensureDir(dir);

		process.chdir(dir);
	});

	afterEach(async () => {
		await del(path.resolve(dir, '**/*'), {
			dot: true,
		});

		process.chdir(cwd);
	});

	afterAll(async () => {
		process.chdir(path.resolve(dir, '../'));

		await del(dir, {
			dot: true,
		});

		process.chdir(cwd);
	});

	test('removes full file tree', async () => {
		const previousFiles = [{ makeDirs: ['dist/static'] }];
		const files: any = [];

		const { parsedFiles, previousStats } = await dirInfo(
			files,
			previousFiles,
		);

		await removeStaleDirectories(parsedFiles, previousStats);

		expect(delMock.mock.calls).toMatchSnapshot();
	});

	test('removes full file tree with multiple nested', async () => {
		const previousFiles = [{ makeDirs: ['dist/static', 'dist/inside'] }];
		const files: ReadonlyArray<void> = [];

		const { parsedFiles, previousStats } = await dirInfo(
			files,
			previousFiles,
		);

		await removeStaleDirectories(parsedFiles, previousStats);

		expect(delMock.mock.calls).toMatchSnapshot();
	});

	test('removes top next directory, but leaves sub managed directory', async () => {
		const previousFiles = [{ makeDirs: ['dist/static'] }];
		const files = [{ makeDirs: ['dist'] }];

		const { parsedFiles, previousStats } = await dirInfo(
			files,
			previousFiles,
		);

		await removeStaleDirectories(parsedFiles, previousStats);

		expect(delMock.mock.calls).toMatchSnapshot();
	});

	test('does not remove base directory when nested directory added', async () => {
		const previousFiles = [{ makeDirs: ['dist'] }];
		const files = [{ makeDirs: ['dist/static'] }];

		const { parsedFiles, previousStats } = await dirInfo(
			files,
			previousFiles,
		);

		await removeStaleDirectories(parsedFiles, previousStats);

		expect(delMock.mock.calls).toEqual([]);
	});

	test('leaves directory if files exist', async () => {
		const previousFiles = [{ makeDirs: ['dist'] }];
		const files: any = [];

		const { parsedFiles, previousStats } = await dirInfo(
			files,
			previousFiles,
		);

		const file = path.resolve(dir, 'dist', 'file1.js');
		await fse.writeJson(file, {});

		await removeStaleDirectories(parsedFiles, previousStats);

		expect(delMock.mock.calls).toEqual([]);
	});

	test('leaves directory if files exist - hidden files', async () => {
		const previousFiles = [{ makeDirs: ['dist'] }];
		const files: any = [];

		const { parsedFiles, previousStats } = await dirInfo(
			files,
			previousFiles,
		);

		const file = path.resolve(dir, 'dist', '.hidden.js');
		await fse.writeJson(file, {});

		await removeStaleDirectories(parsedFiles, previousStats);

		expect(delMock.mock.calls).toEqual([]);
	});

	test('does nothing if directory does not exist but removes root', async () => {
		const previousFiles = [{ makeDirs: ['dist/static'] }];
		const files: any = [];

		const { parsedFiles, previousStats } = await dirInfo(
			files,
			previousFiles,
		);

		const dirPath = path.resolve(dir, 'dist', 'static');
		await del(dirPath);

		await removeStaleDirectories(parsedFiles, previousStats);

		expect(delMock.mock.calls).toMatchSnapshot();
	});
});
