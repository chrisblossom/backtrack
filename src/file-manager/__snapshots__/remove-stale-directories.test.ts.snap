// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeStaleDirectories does nothing if directory does not exist but removes root 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist",
  ],
]
`;

exports[`removeStaleDirectories removes full file tree 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist/static",
  ],
  Array [
    "<PROJECT_ROOT>/dist",
  ],
]
`;

exports[`removeStaleDirectories removes full file tree with multiple nested 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist/inside",
  ],
  Array [
    "<PROJECT_ROOT>/dist/static",
  ],
  Array [
    "<PROJECT_ROOT>/dist",
  ],
]
`;

exports[`removeStaleDirectories removes top next directory, but leaves sub managed directory 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/dist/static",
  ],
]
`;
